File: tessandiocyle_keyx_cpp/src/merkle.cpp
#include "merkle.hpp"
#include "shake.hpp"
#include <cstring>

namespace td18 {

static void hash_leaf(std::array<uint8_t,32>& out, size_t index, int16_t coeff) {
    uint8_t in[10];
    in[0] = static_cast<uint8_t>(index);
    in[1] = static_cast<uint8_t>(index >> 8);
    in[2] = static_cast<uint8_t>(index >> 16);
    in[3] = static_cast<uint8_t>(index >> 24);
    in[4] = static_cast<uint8_t>(index >> 32);
    in[5] = static_cast<uint8_t>(index >> 40);
    in[6] = static_cast<uint8_t>(index >> 48);
    in[7] = static_cast<uint8_t>(index >> 56);
    in[8] = static_cast<uint8_t>(coeff);
    in[9] = static_cast<uint8_t>(coeff >> 8);
    Keccak1600::sha3_256(out.data(), in, sizeof(in));
}

static void hash_node(std::array<uint8_t,32>& out,
                      const std::array<uint8_t,32>& left,
                      const std::array<uint8_t,32>& right) {
    uint8_t in[64];
    std::copy(left.begin(), left.end(), in);
    std::copy(right.begin(), right.end(), in + 32);
    Keccak1600::sha3_256(out.data(), in, sizeof(in));
}

MerkleTree merkle_build_tree(const Poly& coeffs) {
    MerkleTree tree;
    std::vector<std::array<uint8_t,32>> current;
    for (int i = 0; i < N; ++i) {
        std::array<uint8_t,32> h;
        hash_leaf(h, i, coeffs[i]);
        current.push_back(h);
    }
    tree.layers.push_back(current);
    while (current.size() > 1) {
        std::vector<std::array<uint8_t,32>> next;
        for (size_t i = 0; i < current.size(); i += 2) {
            if (i + 1 < current.size()) {
                std::array<uint8_t,32> h;
                hash_node(h, current[i], current[i+1]);
                next.push_back(h);
            } else {
                next.push_back(current[i]);
            }
        }
        current = next;
        tree.layers.push_back(current);
    }
    return tree;
}

MerkleProof merkle_proof(const MerkleTree& tree, size_t index) {
    MerkleProof proof;
    size_t idx = index;
    for (size_t layer = 0; layer + 1 < tree.layers.size(); ++layer) {
        size_t sibling = (idx % 2 == 0) ? (idx + 1) : (idx - 1);
        if (sibling >= tree.layers[layer].size()) sibling = idx;
        proof.push_back({sibling, tree.layers[layer][sibling]});
        idx /= 2;
    }
    return proof;
}

int merkle_verify(const std::array<uint8_t,32>& root, size_t index,
                  int16_t coeff, const MerkleProof& proof) {
    std::array<uint8_t,32> current;
    hash_leaf(current, index, coeff);
    size_t idx = index;
    for (const auto& step : proof) {
        std::array<uint8_t,32> h;
        if (idx % 2 == 0) {
            hash_node(h, current, step.second);
        } else {
            hash_node(h, step.second, current);
        }
        current = h;
        idx /= 2;
    }
    return (current == root) ? 0 : -1;
}

} // namespace td18

File: tessandiocyle_keyx/src/merkle.rs
// TessanDioKey v18 — Merkle Tree Polynomial Commitment for Transparency
//
// During keypair generation, Alice builds a Merkle tree over her secret-key
// polynomial coefficients. She publishes the root hash alongside pk. Later,
// she can reveal individual coefficients (via Merkle paths) for audit or
// verifiable compromise detection without exposing the full secret key.

/// Compute a leaf hash for a single coefficient.
fn leaf_hash(index: usize, coeff: i16) -> [u8; 32] {
    let mut hasher = blake3::Hasher::new();
    hasher.update(&index.to_le_bytes());
    hasher.update(&coeff.to_le_bytes());
    *hasher.finalize().as_bytes()
}

/// Compute the parent hash of two child hashes.
fn node_hash(left: &[u8; 32], right: &[u8; 32]) -> [u8; 32] {
    let mut hasher = blake3::Hasher::new();
    hasher.update(left);
    hasher.update(right);
    *hasher.finalize().as_bytes()
}

/// Build a Merkle tree from a slice of coefficients.
/// Returns the root hash and the tree layers (bottom-up).
pub fn build_merkle_tree(coeffs: &[i16]) -> ([u8; 32], Vec<Vec<[u8; 32]>>) {
    let mut layers: Vec<Vec<[u8; 32]>> = Vec::new();

    // Bottom layer: leaf hashes
    let mut current: Vec<[u8; 32]> = coeffs.iter().enumerate()
        .map(|(i, c)| leaf_hash(i, *c))
        .collect();
    layers.push(current.clone());

    // Build upwards
    while current.len() > 1 {
        let mut next = Vec::new();
        for chunk in current.chunks(2) {
            if chunk.len() == 2 {
                next.push(node_hash(&chunk[0], &chunk[1]));
            } else {
                // Odd node: duplicate itself as sibling
                next.push(node_hash(&chunk[0], &chunk[0]));
            }
        }
        layers.push(next.clone());
        current = next;
    }

    (current[0], layers)
}

/// Generate a Merkle proof for coefficient at `index`.
/// Returns the sibling hashes from leaf to root.
pub fn merkle_proof(tree: &[Vec<[u8; 32]>], index: usize) -> Vec<(usize, [u8; 32])> {
    let mut proof = Vec::new();
    let mut idx = index;

    for layer in tree.iter().take(tree.len() - 1) {
        let sibling_idx = if idx.is_multiple_of(2) { idx + 1 } else { idx - 1 };
        let sibling = if sibling_idx < layer.len() {
            layer[sibling_idx]
        } else {
            // Duplicate current node if no sibling
            layer[idx]
        };
        proof.push((sibling_idx, sibling));
        idx /= 2;
    }

    proof
}

/// Verify a Merkle proof against a root hash.
/// `index` is the coefficient index, `coeff` is the claimed value,
/// `proof` is the list of sibling hashes.
pub fn verify_merkle_proof(root: &[u8; 32], index: usize, coeff: i16, proof: &[(usize, [u8; 32])]) -> bool {
    let mut current = leaf_hash(index, coeff);
    let mut idx = index;

    for (_sibling_idx, sibling_hash) in proof.iter() {
        current = if idx.is_multiple_of(2) {
            node_hash(&current, sibling_hash)
        } else {
            node_hash(sibling_hash, &current)
        };
        idx /= 2;
    }

    &current == root
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_merkle_tree_basic() {
        let coeffs: Vec<i16> = (0..256).map(|i| (i * 13 % 3329) as i16).collect();
        let (root, tree) = build_merkle_tree(&coeffs);
        assert!(!root.iter().all(|&b| b == 0));
        assert!(!tree.is_empty());
    }

    #[test]
    fn test_merkle_proof_verification() {
        let coeffs: Vec<i16> = (0..256).map(|i| (i * 13 % 3329) as i16).collect();
        let (root, tree) = build_merkle_tree(&coeffs);

        for idx in [0usize, 1, 127, 128, 255] {
            let proof = merkle_proof(&tree, idx);
            assert!(verify_merkle_proof(&root, idx, coeffs[idx], &proof));
        }
    }

    #[test]
    fn test_merkle_tamper_detected() {
        let coeffs: Vec<i16> = (0..256).map(|i| (i * 13 % 3329) as i16).collect();
        let (root, tree) = build_merkle_tree(&coeffs);
        let idx = 42;
        let proof = merkle_proof(&tree, idx);

        // Tamper with coefficient
        assert!(!verify_merkle_proof(&root, idx, coeffs[idx] + 1, &proof));
    }

    #[test]
    fn test_merkle_powers_of_two() {
        let coeffs: Vec<i16> = (0..16).map(|i| i as i16).collect();
        let (root, tree) = build_merkle_tree(&coeffs);
        assert_eq!(tree.len(), 5); // 16 → 8 → 4 → 2 → 1

        for (idx, coeff) in coeffs.iter().enumerate().take(16) {
            let proof = merkle_proof(&tree, idx);
            assert!(verify_merkle_proof(&root, idx, *coeff, &proof));
        }
    }
}
